File: _archive/freee-sync-ts/src/db/connection.ts
import Database from 'better-sqlite3';
import path from 'path';
import fs from 'fs';

/**
 * SQLite database connection manager for sync history
 *
 * Purpose:
 * - Manages connection to sync-history.db
 * - Provides query helpers
 * - Ensures WAL mode for better concurrency
 */
export class DatabaseConnection {
  private db: Database.Database;
  private dbPath: string;

  constructor(dbPath: string) {
    this.dbPath = path.resolve(dbPath);

    // Ensure database file exists
    if (!fs.existsSync(this.dbPath)) {
      throw new Error(
        `Database file not found: ${this.dbPath}\n` +
        'Please run: ./scripts/init-db.sh'
      );
    }

    // Open database with better-sqlite3
    this.db = new Database(this.dbPath, {
      verbose: process.env.DEBUG_SQL ? console.log : undefined,
    });

    // Enable WAL mode for better concurrency
    // WAL (Write-Ahead Logging) allows multiple readers while writing
    this.db.pragma('journal_mode = WAL');

    // Enable foreign key constraints
    this.db.pragma('foreign_keys = ON');
  }

  /**
   * Get the underlying database instance
   * Use this for custom queries not covered by sync-history.ts
   */
  getDb(): Database.Database {
    return this.db;
  }

  /**
   * Execute a query and return all results
   */
  query<T = any>(sql: string, params?: any[]): T[] {
    const stmt = this.db.prepare(sql);
    return (params && params.length > 0 ? stmt.all(params) : stmt.all()) as T[];
  }

  /**
   * Execute a query and return the first result
   */
  queryOne<T = any>(sql: string, params?: any[]): T | undefined {
    const stmt = this.db.prepare(sql);
    return (params && params.length > 0 ? stmt.get(params) : stmt.get()) as T | undefined;
  }

  /**
   * Execute an INSERT/UPDATE/DELETE query
   * Returns info about the operation (changes, lastInsertRowid)
   */
  execute(sql: string, params?: any[]): Database.RunResult {
    const stmt = this.db.prepare(sql);
    return params && params.length > 0 ? stmt.run(params) : stmt.run();
  }

  /**
   * Begin a transaction
   * Returns a transaction object with commit/rollback methods
   */
  transaction<T>(fn: () => T): T {
    return this.db.transaction(fn)();
  }

  /**
   * Close the database connection
   */
  close(): void {
    this.db.close();
  }

  /**
   * Check if database is open
   */
  isOpen(): boolean {
    return this.db.open;
  }

  /**
   * Get database path
   */
  getPath(): string {
    return this.dbPath;
  }
}

/**
 * Create a database connection with default path
 */
export function createConnection(dbPath?: string): DatabaseConnection {
  const defaultPath = path.join(process.cwd(), 'data', 'sync-history.db');
  return new DatabaseConnection(dbPath || defaultPath);
}

/**
 * Singleton instance for the default database
 */
let defaultConnection: DatabaseConnection | null = null;

/**
 * Get the default database connection
 * Creates one if it doesn't exist
 */
export function getDefaultConnection(): DatabaseConnection {
  if (!defaultConnection) {
    defaultConnection = createConnection();
  }
  return defaultConnection;
}

/**
 * Close the default database connection
 */
export function closeDefaultConnection(): void {
  if (defaultConnection) {
    defaultConnection.close();
    defaultConnection = null;
  }
}

File: pkg/db/connection.go
package db

import (
	"database/sql"
	"fmt"
	"os"
	"path/filepath"

	_ "github.com/mattn/go-sqlite3" // SQLite driver
)

// Connection manages a SQLite database connection.
type Connection struct {
	db     *sql.DB
	dbPath string
}

// Open opens a SQLite database connection.
// It enables WAL mode for better concurrency and foreign key constraints.
func Open(dbPath string) (*Connection, error) {
	// Ensure database file's parent directory exists
	dbDir := filepath.Dir(dbPath)
	if err := os.MkdirAll(dbDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create database directory: %w", err)
	}

	// Open database with SQLite driver
	// Connection string enables foreign keys and WAL mode
	connStr := fmt.Sprintf("file:%s?_foreign_keys=on&_journal_mode=WAL", dbPath)
	db, err := sql.Open("sqlite3", connStr)
	if err != nil {
		return nil, fmt.Errorf("failed to open database: %w", err)
	}

	// Test connection
	if err := db.Ping(); err != nil {
		db.Close()
		return nil, fmt.Errorf("failed to ping database: %w", err)
	}

	conn := &Connection{
		db:     db,
		dbPath: dbPath,
	}

	// Initialize schema
	if err := InitializeSchema(conn); err != nil {
		db.Close()
		return nil, fmt.Errorf("failed to initialize schema: %w", err)
	}

	return conn, nil
}

// Close closes the database connection.
func (c *Connection) Close() error {
	if c.db != nil {
		return c.db.Close()
	}
	return nil
}

// GetDB returns the underlying *sql.DB instance.
// Use this for custom queries not covered by other methods.
func (c *Connection) GetDB() *sql.DB {
	return c.db
}

// GetPath returns the database file path.
func (c *Connection) GetPath() string {
	return c.dbPath
}

// Query executes a query that returns rows.
func (c *Connection) Query(query string, args ...interface{}) (*sql.Rows, error) {
	return c.db.Query(query, args...)
}

// QueryRow executes a query that is expected to return at most one row.
func (c *Connection) QueryRow(query string, args ...interface{}) *sql.Row {
	return c.db.QueryRow(query, args...)
}

// Exec executes a query that doesn't return rows.
// Returns sql.Result with information about the operation (LastInsertId, RowsAffected).
func (c *Connection) Exec(query string, args ...interface{}) (sql.Result, error) {
	return c.db.Exec(query, args...)
}

// Begin starts a new transaction.
func (c *Connection) Begin() (*sql.Tx, error) {
	return c.db.Begin()
}

// Transaction executes a function within a transaction.
// If the function returns an error, the transaction is rolled back.
// Otherwise, the transaction is committed.
func (c *Connection) Transaction(fn func(*sql.Tx) error) error {
	tx, err := c.Begin()
	if err != nil {
		return fmt.Errorf("failed to begin transaction: %w", err)
	}

	defer func() {
		if p := recover(); p != nil {
			tx.Rollback()
			panic(p)
		}
	}()

	if err := fn(tx); err != nil {
		if rbErr := tx.Rollback(); rbErr != nil {
			return fmt.Errorf("transaction error: %v, rollback error: %w", err, rbErr)
		}
		return err
	}

	if err := tx.Commit(); err != nil {
		return fmt.Errorf("failed to commit transaction: %w", err)
	}

	return nil
}
